jxl/src/container/parse.rs

```diff
@@ -0,0 +1,269 @@
+                }
+                DetectState::InCodestream {
+                    bytes_left: Some(bytes_left),
+                    ..
+                } => {
+                    let payload = if buf.len() >= *bytes_left {
+                        let (payload, remaining) = buf.split_at(*bytes_left);
+                        *state = DetectState::WaitingBoxHeader;
+                        *buf = remaining;
+                        payload
+                    } else {
+                        let payload = *buf;
+                        *bytes_left -= buf.len();
+                        *buf = &[];
+                        payload
+                    };
+                    return Ok(Some(ParseEvent::Codestream(payload)));
+                }
+                DetectState::InAuxBox {
+                    header: _,
+                    bytes_left: None,
+                } => {
+                    let _payload = *buf;
+                    *buf = &[];
+                    // FIXME: emit auxiliary box event
+                }
+                DetectState::InAuxBox {
+                    header: _,
+                    bytes_left: Some(bytes_left),
+                } => {
+                    let _payload = if buf.len() >= *bytes_left {
+                        let (payload, remaining) = buf.split_at(*bytes_left);
+                        *state = DetectState::WaitingBoxHeader;
+                        *buf = remaining;
+                        payload
+                    } else {
+                        let payload = *buf;
+                        *bytes_left -= buf.len();
+                        *buf = &[];
+                        payload
+                    };
+                    // FIXME: emit auxiliary box event
+                }
+            }
+        }
+    }
+}
+
+impl std::fmt::Debug for ParseEvents<'_, '_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("ParseEvents")
+            .field("inner", &self.inner)
+            .field(
+                "remaining_input",
+                &format_args!("({} byte(s))", self.remaining_input.len()),
+            )
+            .field("finished", &self.finished)
+            .finish()
+    }
+}
+
+impl<'inner, 'buf> Iterator for ParseEvents<'inner, 'buf> {
+    type Item = Result<ParseEvent<'buf>>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.finished {
+            return None;
+        }
+
+        let initial_buf = self.remaining_input;
+        let event = self.emit_single();
+
+        if event.is_err() {
+            self.finished = true;
+        }
+
+        self.inner.previous_consumed_bytes += initial_buf.len() - self.remaining_input.len();
+        event.transpose()
+    }
+}
+
+/// Parser event emitted by [`ParseEvents`].
+pub enum ParseEvent<'buf> {
+    /// Bitstream structure is detected.
+    BitstreamKind(BitstreamKind),
+    /// Codestream data is read.
+    Codestream(&'buf [u8]),
```

Comment:
if I understand correctly, this need not be the full codestream, is that correct?
If so, maybe add a comment about that?